### PR TITLE
feat(bot,rest,types,utils): Support User-Installed apps, Interaction contexts and install types

### LINKED ISSUE #3492: [api-docs] Add docs for user-installed apps (#6723)
A new commit was made into the api-docs repo: https://github.com/discord/discord-api-docs/commit/1eaa1af5677712957479c9959132003b79396e24
add user install docs + new pages for interactions and overview of apps

This is a bot created issue.

### LINKED ISSUE #3495: [api-docs] fix optionality for global commands (#6726)
A new commit was made into the api-docs repo: https://github.com/discord/discord-api-docs/commit/fc56b3d618259a326607d52de00c47e6fd1112b7

This is a bot created issue.

### LINKED ISSUE #3496: [api-docs] Add integration_types_config to params of Application Edit endpoint (#6727)
A new commit was made into the api-docs repo: https://github.com/discord/discord-api-docs/commit/72197509c5bd7db4a3377419885b3fed98506d4c

This is a bot created issue.

### LINKED ISSUE #3498: [api-docs] clarify to command create (#6728)
A new commit was made into the api-docs repo: https://github.com/discord/discord-api-docs/commit/00a67f7b3922f3547e791414c6326aa2ae37644d

This is a bot created issue.

### LINKED ISSUE #3517: [api-docs] Document `integration_type` oauth parameter (#6738)
A new commit was made into the api-docs repo: https://github.com/discord/discord-api-docs/commit/723fabf9a0762e20b6a4204c23d9805fb28ef39a
* Document `integration_type` oauth parameter

* Update OAuth2.md

---------

Co-authored-by: Yoni Lerner <yonilerner@gmail.com>

This is a bot created issue.

### LINKED ISSUE #3523: [api-docs] App command integration_types/contexts nullability (#6765)
A new commit was made into the api-docs repo: https://github.com/discord/discord-api-docs/commit/8396be30d8b2961e4a1bfd3ad3f9d92821d7aeca
* App command integration_types/contexts nullability

* fix: wrong table row (#6766)

* fix: wrong table row

* yet another fix

---------

Co-authored-by: Lala Sabathil <aiko@aitsys.dev>

This is a bot created issue.

### LINKED ISSUE #3525: [api-docs] User install edits (#6782)
A new commit was made into the api-docs repo: https://github.com/discord/discord-api-docs/commit/a470279caba02de43de780e09ba3dd54b4b6220e
* oauth2_install_params optional

* wording

* add preview language

* tables

This is a bot created issue.

### PR DESCRIPTION
**Pull request by Fleny113**
- Support User-Installed applications
- Support to choose if the command is for a user-based install or guild-based install
- Support for contexts
- Add support for the new integrationTypesConfig option in the application

These changes deprecated some existing stuff. Such as
- `CreateSlashApplicationCommand.dmPermission` in favor of using `CreateSlashApplicationCommand.contexts`
- `DiscordMessage.interaction` (and the transformed counterpart) in favor of `DiscordMessage.interaction_metadata`

When reviewing this pr i suggest checking the discord api docs commits too as I tried to update everything but i might have got something wrong or overlooked something.

closes #3496
closes #3495
closes #3498
closes #3492
closes #3517
closes #3523
closes #3525

### CODE REVIEW

File: packages/types/src/shared.ts [hunk 1]
```diff
@@ -1024,7 +1024,7 @@ export type Localization = Partial<Record<Locales, string>>
 
 export type AtLeastOne<T, U = { [K in keyof T]: Pick<T, K> }> = Partial<T> & U[keyof U]
 export type CamelCase<S extends string> = S extends `${infer T}_${infer U}` ? `${T}${Capitalize<CamelCase<U>>}` : S
-export type SnakeCase<S extends string> = S extends `${infer T}${infer U}` ? `${T extends Capitalize<T> ? '_' : ''}${Lowercase<T>}${SnakeCase<U>}` : S
+export type SnakeCase<S extends string> = S extends `${infer T}${infer U}` ? `${T extends Lowercase<T> ? '' : '_'}${Lowercase<T>}${SnakeCase<U>}` : S
```
Comment (lts20050703) [inline packages/types/src/shared.ts:1027]: Unrelated change?

Comment (Fleny113) [reply packages/types/src/shared.ts:1027]: Looks like one. It is not. It was breaking stuff bc discord had the great of using numbers in JSON keys or something (there is a discord message I asked help on this: https://discord.com/channels/785384884197392384/785384884197392387/1219773105493377104)

**Review by lts20050703 — APPROVED** [review]

**Review by MatthewSH — APPROVED** [review]

**Review by MatthewSH — APPROVED** [review]

**Review by lts20050703 — APPROVED** [review]
